pkg/chunk/cassandra/storage_client.go

```diff
@@ -0,0 +1,219 @@
+	schemaCfg chunk.SchemaConfig
+	session   *gocql.Session
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewStorageClient(cfg Config, schemaCfg chunk.SchemaConfig) (chunk.StorageClient, error) {
+	session, err := cfg.session()
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageClient{
+		cfg:       cfg,
+		schemaCfg: schemaCfg,
+		session:   session,
+	}, nil
+}
+
+func (s *storageClient) Close() {
+	s.session.Close()
+}
+
+type writeBatch struct {
+	b *gocql.Batch
+}
+
+func (s *storageClient) NewWriteBatch() chunk.WriteBatch {
+	return writeBatch{
+		b: gocql.NewBatch(gocql.UnloggedBatch),
+	}
+}
+
+func (b writeBatch) Add(tableName, hashValue string, rangeValue []byte, value []byte) {
+	b.b.Query(fmt.Sprintf("INSERT INTO %s (hash, range, value) VALUES (?, ?, ?)", tableName),
+		hashValue, rangeValue, value)
+}
+
+func (s *storageClient) BatchWrite(ctx context.Context, batch chunk.WriteBatch) error {
+	cassandraBatch := batch.(writeBatch)
+	return s.session.ExecuteBatch(cassandraBatch.b.WithContext(ctx))
+}
+
+func (s *storageClient) QueryPages(ctx context.Context, query chunk.IndexQuery, callback func(result chunk.ReadBatch, lastPage bool) (shouldContinue bool)) error {
+	var q *gocql.Query
+	if len(query.RangeValuePrefix) > 0 {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ? AND range >= ?", query.TableName),
+			query.HashValue, query.RangeValuePrefix)
+	} else if len(query.RangeValueStart) > 0 {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ? AND range >= ?", query.TableName),
+			query.HashValue, query.RangeValueStart)
+	} else {
+		q = s.session.Query(fmt.Sprintf("SELECT range, value FROM %s WHERE hash = ?", query.TableName),
+			query.HashValue)
+	}
+
+	iter := q.WithContext(ctx).Iter()
+	defer iter.Close()
+	scanner := iter.Scanner()
+	for scanner.Next() {
+		var b readBatch
+		if err := scanner.Scan(&b.rangeValue, &b.value); err != nil {
+			return err
+		}
+		if callback(b, false) {
+			return nil
+		}
+	}
+	return scanner.Err()
+}
+
+// readBatch represents a batch of rows read from Cassandre.
+type readBatch struct {
+	rangeValue []byte
+	value      []byte
+}
+
+func (readBatch) Len() int {
+	return 1
+}
+
+func (b readBatch) RangeValue(index int) []byte {
+	if index != 0 {
+		panic("index != 0")
+	}
+	return b.rangeValue
+}
+
+func (b readBatch) Value(index int) []byte {
+	if index != 0 {
+		panic("index != 0")
+	}
+	return b.value
+}
+
+func (s *storageClient) PutChunks(ctx context.Context, chunks []chunk.Chunk) error {
+	b := gocql.NewBatch(gocql.UnloggedBatch).WithContext(ctx)
+
+	for i := range chunks {
+		// Encode the chunk first - checksum is calculated as a side effect.
+		buf, err := chunks[i].Encode()
+		if err != nil {
+			return err
+		}
+		key := chunks[i].ExternalKey()
+		tableName := s.schemaCfg.ChunkTables.TableFor(chunks[i].From)
+		b.Query(fmt.Sprintf("INSERT INTO %s (hash, range, value) VALUES (?, 0x00, ?)", tableName), key, buf)
```

Comment:
Nit/Magic value: a comment about why `0x00` is so special might help the next reader/maintainer.